### PR TITLE
Invocation retry

### PR DESCRIPTION
**Pull request by GhostInAMachine**
You can test retrying by setting up a map interceptor which would throw an exception like NullPointerException and then adding a relevant exception code to isRetryable method in RemoteException class

### CODE REVIEW

File: src/invocation/InvocationService.ts [hunk 1]
```diff
@@ -106,19 +124,35 @@ export class InvocationService {
         var messageType = clientMessage.getMessageType();
 
         if (clientMessage.hasFlags(BitsUtil.LISTENER_FLAG)) {
-            Q.fcall(this.eventHandlers[correlationId].handler, clientMessage);
```
Comment (mustafaiman) [inline src/invocation/InvocationService.ts]: I put there a fcall because it puts the handler at the end of the event loop. If we call the handler directly, it is called immediately. We should break execution logic into parts as best as we can. You may use `setImmediate` instead of `Q.fcall` if unused promise bothers you.

File: src/invocation/InvocationService.ts [hunk 2]
```diff
@@ -8,19 +8,27 @@ import Address = require('../Address');
 import ExceptionCodec = require('../codec/ExceptionCodec');
 import {BitsUtil} from '../BitsUtil';
 
-export interface Invocation {
+var EXCEPTION_MESSAGE_TYPE = 109;
+var INVOCATION_TIMEOUT = 120000;
+var INVOCATION_RETRY_DELAY = 1000;
+
+class Invocation {
```
Comment (mustafaiman) [inline src/invocation/InvocationService.ts]: If we do not `export Invocation` no other class can create invocation by itself and call `InvocationService.invoke` directly. Do we want that? I think `Invocation` should be exported.

File: src/invocation/InvocationService.ts [hunk 3]
```diff
@@ -8,19 +8,27 @@ import Address = require('../Address');
...
+
     request: ClientMessage;
-    partitionId?: number;
-    address?: Address;
-    connection?: ClientConnection;
-    deferred?: Q.Deferred<ClientMessage>;
-    handler?: (...args: any[]) => any;
+    partitionId: number;
+    address: Address;
+    deadline: Date = new Date(new Date().getTime() + INVOCATION_TIMEOUT);
```
Comment (mustafaiman) [inline src/invocation/InvocationService.ts]: I see we retry if an exception comes from server but how do we handle the situation where no response comes at all? I did not see any code handling such situation.

Comment (GhostInAMachine) [inline src/invocation/InvocationService.ts]: Timing out will be added in a different PR

File: src/invocation/InvocationService.ts (hunk 2 above)
Comment (GhostInAMachine) [inline src/invocation/InvocationService.ts]: Why should any other component have a direct access to invoke?

Comment (mustafaiman) [inline src/invocation/InvocationService.ts]: Maybe it is not needed. Ok let's leave it unexported.

File: src/invocation/InvocationService.ts (hunk 3 above)
Comment (mustafaiman) [inline src/invocation/InvocationService.ts]: Ok. I created an issue so we do not forget it. https://github.com/hazelcast/hazelcast-nodejs-client/issues/52

File: src/invocation/InvocationService.ts (hunk 1 above)
Comment (GhostInAMachine) [inline src/invocation/InvocationService.ts]: @mustafaiman Latest commit wraps the callback with setImmediate

**Comment by mustafaiman** [comment]
:+1: Rebase please
